### ISSUE TITLE
dataflow machine simulator

### ISSUE THREAD

**Issue by yegor256** [objectionary/ideas#44]
_Dataflow architecture_ is a computer architecture that directly contrasts the traditional von Neumann architecture or control flow architecture. The latter executes instructions sequentially, while the former doesn't have a program counter. A dataflow architecture represents a program as a _dataflow graph_ (DFG), where nodes are "operators" and directed edges between them are the "channels" through which immutable data values travel. An operator is evaluated when data are present on all input channels. The completion of evaluation makes the resulting data available to the operators at the ends of output channels. Thus, the order of operator evaluation is unpredictable, i.e., behavior is _nondeterministic_.

In order to study the behavior of a DFG we can create a simulator of a dataflow machine. It should be a command line tool. I should be able to specify the DFG (in JS-like language) in a file and then run it:

```
$ cat foo.dfg
function v1(input) { return { z : input.x + input.y }; }
function v2(input) { console.log(input.a); return { }; }
"42" -> v1.x
"77" -> v1.y
v1.z -> v2.a
```

The output should look like this:

```
$ simulator foo.dfg
v1.x : "42"
v1.y : "77"
run(v1)
v2.a : "119"
```

Simply put, the output should show step by step which data show up on each edge.

**Comment by yegor256** [comment]
a prototype is here: https://github.com/yegor256/damsi
